rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java

```diff
@@ -1066,4 +1103,73 @@ public Response index() throws URISyntaxException {
         return Response.seeOther(new URI("doc/jaxrsdocs/rm/index.html")).build();
     }
 
+    private void prepareInfraParams(@FormParam("infrastructureType") String infrastructureType,
+            @FormParam("infrastructureParameters") String[] infrastructureParameters,
+            @FormParam("infrastructureFileParameters") String[] infrastructureFileParameters, ResourceManager rm,
+            Object[] infraParams) {
+        /*
+         * we need to merge both infrastructureParameters and
+         * infrastructureFileParameters into one to do so we need the
+         * infrastructure parameter order from the RM
+         */
+        for (PluginDescriptor infra : rm.getSupportedNodeSourceInfrastructures()) {
+            if (infra.getPluginName().equals(infrastructureType)) {
+                assignInfraFieldToNormalParamOrToFileParam(infrastructureParameters,
+                                                           infrastructureFileParameters,
+                                                           infraParams,
+                                                           infra);
+            }
+        }
+    }
+
+    private void assignInfraFieldToNormalParamOrToFileParam(
+            @FormParam("infrastructureParameters") String[] infrastructureParameters,
+            @FormParam("infrastructureFileParameters") String[] infrastructureFileParameters, Object[] infraParams,
+            PluginDescriptor infra) {
+        int i = 0, j = 0, k = 0;
+        for (ConfigurableField field : infra.getConfigurableFields()) {
+            if (field.getMeta().credential() || field.getMeta().fileBrowser()) {
+                /*
+                 * file parameter : insert from the other array, convert
+                 * to byte[]
+                 */
+                infraParams[i] = infrastructureFileParameters[k].getBytes();
+                k++;
+            } else {
+                infraParams[i] = infrastructureParameters[j];
+                j++;
+            }
+            i++;
+        }
+    }
+
+    private void preparePolicyParams(@FormParam("policyType") String policyType,
+            @FormParam("policyParameters") String[] policyParameters,
+            @FormParam("policyFileParameters") String[] policyFileParameters, ResourceManager rm,
+            Object[] policyParams) {
+        for (PluginDescriptor pol : rm.getSupportedNodeSourcePolicies()) {
+            if (pol.getPluginName().equals(policyType)) {
+                assignPolicyFieldToNormalParamOrToFileParam(policyParameters, policyFileParameters, policyParams, pol);
+            }
+        }
+    }
+
+    private void assignPolicyFieldToNormalParamOrToFileParam(@FormParam("policyParameters") String[] policyParameters,
+            @FormParam("policyFileParameters") String[] policyFileParameters, Object[] policyParams,
+            PluginDescriptor pol) {
+        int i = 0, j = 0, k = 0;
```

Comment:
The code flow is very hard to follow, with the three indexes. Is it possible to re-write this. 

What I understand is that you iterate through the configurable fields and if the field fulfills specific properties then you save an argument from the method into the running index of policyParams.

Can you somehow express the releationship between pol and policyFileParameters, policyParameters. It is not clear to me why you can iterate through pol and then save policyFileParameters or policyParameters without checking their length.

Reply:
You could rename k to policyFileParametersIndex and j to policyParametersIndex.

Reply:
The truth is that I moved this code to a new method, but it has been there for ages. I am not confident at all rewriting it in a way that would use less indexing and tricks.
However what I will do as you suggested is to rename as accurately as possible the indexes and variables.